### PR TITLE
[KEH-413] - Terraform API

### PR DESCRIPTION
**Pull request by evisos-ons**
# Terraform Infrastructure Management

This repository contains Terraform configurations for managing various AWS services including storage, Lambda functions, authentication, API Gateway and CloudWatch.

### Prerequisites
- Terraform installed (version >= 4.0.0)
- AWS credentials configured
- Appropriate `.tfvars` and `.tfbackend` files in the `env/dev/` directory

### Directory Structure
The infrastructure is organized into separate services:
- `terraform/secrets/` - Create Secrets Manager to store variables
- `terraform/storage/` - S3 buckets and DynamoDB tables
- `terraform/lambda/` - Lambda functions
- `terraform/authentication/` - Cognito user pools
- `terraform/api_gateway/` - API Gateway configuration

### Common Commands

```bash
cd terraform/<storage/lambda/authentication/api_gateway> 

terraform init -backend-config=env/dev/backend-dev.tfbackend -reconfigure

terraform refresh -var-file=env/dev/dev.tfvars

terraform validate

terraform plan -var-file=env/dev/dev.tfvars

terraform apply -var-file=env/dev/dev.tfvars
```

### CODE REVIEW

File: terraform/lambda/env/dev/backend-dev.tfbackend
Comment (TotalDwarf03) [inline terraform/lambda/env/dev/backend-dev.tfbackend:1]: As above

File: terraform/secrets/env/dev/backend-dev.tfbackend
Comment (TotalDwarf03) [inline terraform/secrets/env/dev/backend-dev.tfbackend:1]: As above

File: terraform/secrets/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+Mandatory variables:
+```
+aws_account_id        = "<ACCOUNT_ID>"
+aws_access_key_id     = "<ACCESS_KEY_ID>"
+aws_secret_access_key = "<SECRET_ACCESS_KEY>"
+domain                = "sdp-dev"
+```
+
+Optional variables that can be set later:
```
Comment (TotalDwarf03) [inline terraform/secrets/README.md:11]: Can we just include these as mandatory? They need to be set for the app to work no?

File: terraform/secrets/env/dev/example_tfvars.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+aws_account_id        = "<your_account_id>"
```
Comment (TotalDwarf03) [inline terraform/secrets/env/dev/example_tfvars.txt:1]: If the optionals become mandatory, make sure to add them to the example.

File: terraform/lambda/main.tf [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        ]
+        Resource = "*"
+      }
+    ]
+  })
+  depends_on = [aws_iam_role.lambda_execution_role]
+}
+
+# 6. Create the Lambda function
+resource "aws_lambda_function" "tech_audit_lambda" {
```
Comment (TotalDwarf03) [inline terraform/lambda/main.tf:158]: This needs to sit in the private VPC I believe.

@gibbardsteve  can you confirm?

If you take a look at the Policy and CoPilot Lambdas, the logic is there.

https://github.com/ONS-Innovation/github-copilot-usage-dashboard/blob/86f06704e1bc45ea45e4e4287eece4aa75d5c4d5/terraform/data_logger/main.tf#L25

File: terraform/storage/README.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Example sandbox.tfvars file:
```
Comment (TotalDwarf03) [inline terraform/storage/README.md:1]: Does this need to be here if `/env/dev/example_tfvars.txt`?

**Review by TotalDwarf03 — COMMENTED** [review]
Just a couple of minors.

File: terraform/README.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+Once you have setup all the services, you must configure the redirect URLs in the Cognito user pool to point to the API Gateway endpoint.
+
+Go to your [user pools](https://eu-west-2.console.aws.amazon.com/cognito/v2/idp/user-pools?region=eu-west-2) and click on the user pool you created.
+
+Follow these steps:
+
+1. Click on the `App Integrations` tab.
+2. Scroll down to `App clients and analytics` and click on your App client.
+3. Scroll down to `Hosted UI` and click on the `Edit` button.
+4. Click `Add another URL` and add the redirect URLs to the `Callback URLs`. These will be the URLs of the API Gateway endpoint. For example:
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/README.md:48]: Yeah that addresses that. For the time being I added a guide for it.

File: terraform/authentication/README.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+Example sandbox.tfvars file:
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/authentication/README.md:1]: Removed

File: terraform/storage/main.tf
Comment (gibbardsteve) [inline terraform/storage/main.tf:1]: There seem to be two S3 buckets in sdp dev related to the API? Can you clean this up please.

image

File: terraform/storage/main.tf [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+# Remove the ECR repository creation since it already exists
+# Instead, use data source to reference existing ECR
+data "aws_ecr_repository" "tech_audit_tool" {
+  name = var.ecr_repository_name
+}
+
+# Initial ECR repository policy (without Lambda role)
+resource "aws_ecr_repository_policy" "tech_audit_tool_policy" {
```
Comment (gibbardsteve) [inline terraform/storage/main.tf:20]: Why is this needed? The terraform doesn't publish the container to ECR does it?

File: terraform/storage/providers.tf [hunk 8]
```diff
@@ -0,0 +1,23 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 4.0"
```
Comment (gibbardsteve) [inline terraform/storage/providers.tf:5]: 5.0 ?

File: terraform/storage/variables.tf [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  default     = "v0.0.1"
+}
+
+variable "lambda_name" {
+  description = "Name of the Lambda function (without environment prefix)"
+  type        = string
+  default     = "tech-audit-tool-api"
+}
+
+variable "ecr_repository_name" {
```
Comment (gibbardsteve) [inline terraform/storage/variables.tf:64]: If this is the same as the lambda name you could reuse

**Review by gibbardsteve — COMMENTED** [review]
A couple of comments.

File: terraform/README.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+terraform validate
+
+terraform plan -var-file=env/sandbox/sandbox.tfvars
+
+terraform apply -var-file=env/sandbox/sandbox.tfvars
+```
+
+### Authentication
+
+You must create the Cognito user pool (authentication) then create the API Gateway (api_gateway) then define the `cognito_user_pool_arn` and `lambda_function_invoke_arn` variables in the `api_gateway` directory.
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/README.md:37]: Yes. I couldn't find a better way to do it other than combining authentication and API gateway terraform.

File: terraform/README.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+Once you have setup all the services, you must configure the redirect URLs in the Cognito user pool to point to the API Gateway endpoint.
+
+Go to your [user pools](https://eu-west-2.console.aws.amazon.com/cognito/v2/idp/user-pools?region=eu-west-2) and click on the user pool you created.
+
+Follow these steps:
+
+1. Click on the `App Integrations` tab.
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/README.md:45]: I forgot to remove this from the REAME. 

I made a PR for custom token lifespan and custom callback urls [here](https://github.com/ONS-Innovation/keh-cognito-auth-tf-module/pull/2).

File: terraform/lambda/variables.tf [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+  default     = "v0.0.2"
+}
+
+variable "tech_audit_data_bucket_name" {
+  description = "Name of the S3 bucket for tech audit data"
+  type        = string
+  default     = "tech-audit-tool-data"
+}
+
+variable "tech_audit_secret_manager_name" {
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/lambda/variables.tf:58]: It stores the Cognito pool id, client id, client secret and redirect uri in the secret manager.

File: aws_lambda_script/old_dockerfile [hunk 13]
```diff
@@ -0,0 +1,27 @@
+# Use an official Python runtime as a parent image
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply aws_lambda_script/old_dockerfile]: Removed.

File: terraform/api_gateway/env/dev/backend-dev.tfbackend
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/api_gateway/env/dev/backend-dev.tfbackend:1]: Done.

File: aws_lambda_script/Dockerfile [hunk 14]
```diff
@@ -17,11 +13,11 @@ RUN pip install --no-cache-dir -r requirements.txt
...
 
-# Change ownership of the application files to the non-root user
-RUN chown -R appuser:appuser /var/task
+ENV TECH_AUDIT_DATA_BUCKET=sdp-dev-tech-audit-tool-api
+ENV TECH_AUDIT_SECRET_MANAGER=sdp-dev-tech-audit-tool-api/secrets
+ENV AWS_DEFAULT_REGION=eu-west-2
+ENV AWS_COGNITO_TOKEN_URL=https://tech-audit-tool-api-sdp-dev.auth.eu-west-2.amazoncognito.com/oauth2/token
 
-# Switch to non-root user
-USER appuser
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply aws_lambda_script/Dockerfile:25]: Now runs as non-root.

File: terraform/storage/main.tf
Comment (evisos-ons) [reply terraform/storage/main.tf:1]: Removed.

File: terraform/storage/providers.tf (hunk 8 above)
Comment (evisos-ons) [reply terraform/storage/providers.tf]: I had problems with 5.0 when I was building Terraform but I've changed back now.

File: terraform/lambda/main.tf [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+  function_name = "${var.domain}-${var.service_subdomain}-lambda"
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/lambda/main.tf]: Done.

File: terraform/storage/main.tf (hunk 7 above)
Comment (evisos-ons) [reply terraform/storage/main.tf]: To be honest, there was a reason why I added this, however it was the other week so I have to refresh my memory as to why I did. I do remember that it was needed for some reason.

I have removed it though.

File: terraform/storage/variables.tf (hunk 9 above)
Comment (evisos-ons) [reply terraform/storage/variables.tf]: Two comments above removes the need for this variable.

File: terraform/api_gateway/env/dev/example_tfvars.txt [hunk 16]
```diff
@@ -0,0 +1,7 @@
+aws_account_id        = "<your_account_id>"
+aws_access_key_id     = "<your_access_key_id>"
+aws_secret_access_key = "<your_secret_access_key>"
+domain                = "sdp-dev"
+cognito_user_pool_arn = "arn:aws:cognito-idp:REGION:ACCOUNT_ID:userpool/USER_POOL_ID"
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/api_gateway/env/dev/example_tfvars.txt]: Addressed [here](https://jira.ons.gov.uk/secure/RapidBoard.jspa?rapidView=4108&projectKey=KEH&view=detail&selectedIssue=KEH-505) and updated.

File: terraform/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/README.md]: Addressed [here](https://jira.ons.gov.uk/secure/RapidBoard.jspa?rapidView=4108&projectKey=KEH&view=detail&selectedIssue=KEH-505) and updated.

File: terraform/lambda/README.md [hunk 17]
```diff
@@ -0,0 +1,11 @@
+Example sandbox.tfvars file:
+
+```
+aws_account_id                    = "<your_account_id>"
+aws_access_key_id                 = "<your_access_key_id>"
+aws_secret_access_key             = "<your_secret_access_key>"
+domain                            = "sdp-sandbox"
+ecr_repository                    = "keh-tech-audit-tool-lambda-api" # ECR repository name where the lambda image is stored
+tech_audit_data_bucket_name       = "keh-tech-audit-tool" # S3 bucket name where the tech audit data is stored
```
Comment (gibbardsteve) [inline terraform/lambda/README.md:9]: Do these names start with keh or sdp-dev - your example_tfvars is sdp-dev ?

File: terraform/lambda/main.tf (hunk 3 above)
Comment (gibbardsteve) [reply terraform/lambda/main.tf:151]: Yes - it would be better if we run everything that is not public facing in the private subnet of the VPC.

**Review by gibbardsteve — COMMENTED** [review]
A nit and a response to Kieran's comment

File: terraform/lambda/README.md (hunk 17 above)
Comment (evisos-ons) [reply terraform/lambda/README.md:9]: We don't need this README as we have the example_tfvars, so I removed it.

File: terraform/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/README.md]: Uses data.tf

File: terraform/lambda/main.tf (hunk 3 above)
Comment (evisos-ons) [reply terraform/lambda/main.tf:173]: Done [here](https://github.com/ONS-Innovation/keh-tech-audit-tool-api/pull/3/commits/60d69908e1a4c043007f404067677a14857ef7e4)

File: terraform/lambda/variables.tf [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+variable "business_owner_tag" {
+  description = "Business Owner"
+  type        = string
+  default     = "DST"
+}
+
+
+variable "ecr_repository_name" {
```
Comment (TotalDwarf03) [inline terraform/lambda/variables.tf:71]: duplicate of `ecr_repository`

File: terraform/storage/main.tf [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+    # Backend is selected using terraform init -backend-config=path/to/backend-<env>.tfbackend
+    # bucket         = "sdp-dev-tf-state"
+    # key            = "sdp-sandbox-tech-audit-tool-api-lambda/terraform.tfstate"
+    # region         = "eu-west-2"
+    # dynamodb_table = "terraform-state-lock"
+  }
+
+}
+
+# Remove the ECR repository creation since it already exists
```
Comment (TotalDwarf03) [inline terraform/storage/main.tf:13]: Reiterating what Steve said, why is the ECR stuff needed here?

File: terraform/storage/variables.tf [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+  default     = "Knowledge Exchange Hub"
+}
+
+variable "business_owner_tag" {
+  description = "Business Owner"
+  type        = string
+  default     = "DST"
+}
+
+variable "container_ver" {
```
Comment (TotalDwarf03) [inline terraform/storage/variables.tf:52]: Are the ECR variables needed here if this terraform is just making a bucket?

I know the vars are used in `main.tf` but if only resourcing storage, why is ECR needed?

**Review by TotalDwarf03 — COMMENTED** [review]
Few comments to look at before this can be merged.

File: README.md [hunk 21]
```diff
@@ -87,6 +87,20 @@ Retrieve your `id_token`, go to the URL above, click the green outlined button w
 
 Now you can go through the /api/v1/ routes and test the endpoints.
 
+## MkDocs
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply README.md:90]: Going to remove MkDocs stuff as that was in a separate PR, my bad.

File: README.md [hunk 22]
```diff
@@ -207,7 +221,7 @@ Send JSON in this format:
     "stage":"Development"
   }
 ```
-Create's a project. If the languages, database, frameworks, CICD, infrastructure or source control, is not in the array_data.json bucket then it is added.
+Create's a project. If the languages, database, frameworks, cicd, infrastructure or source control, is not in the array_data.json bucket then it is added.
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply README.md]: Done 👍

File: README.md [hunk 23]
```diff
@@ -312,7 +326,7 @@ Send JSON in this format:
     "stage":"Development"
   }
 ```
-Edits a project. If the languages, database, frameworks, CICD, infrastructure or source control, is not in the array_data.json bucket then it is added.
+Edits a project. If the languages, database, frameworks, cicd, infrastructure or source control, is not in the array_data.json bucket then it is added.
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply README.md]: Done 👍

File: aws_lambda_script/Dockerfile [hunk 24]
```diff
@@ -1,5 +1,5 @@
 # Use an official Python runtime as a parent image
-FROM public.ecr.aws/lambda/python:3.12.4
+FROM python:3.12-slim
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply aws_lambda_script/Dockerfile]: Changed back 👍

File: terraform/README.md [hunk 25]
```diff
@@ -0,0 +1,44 @@
+# Terraform Infrastructure Management
+
+This repository contains Terraform configurations for managing various AWS services including storage, Lambda functions, authentication, API Gateway and CloudWatch.
+
+## Getting Started
+
+### Prerequisites
+- Terraform installed (version >= 1.0.0)
+- AWS credentials configured
+- Appropriate `.tfvars` and `.tfbackend` files in the `env/sandbox/` directory
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/README.md]: Done 👍

File: terraform/README.md [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+terraform plan -var-file=env/dev/dev.tfvars
+
+terraform apply -var-file=env/dev/dev.tfvars
+```
+
+### Authentication
+
+You must create the Cognito user pool (authentication) then create the API Gateway (api_gateway).
+
+Once API Gateway is created, you can configure the callback URLs in the Cognito user pool to point to the API Gateway endpoint.
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/README.md]: Added. Also linked the GH pages.

File: terraform/api_gateway/README.md [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+```
+
+`rest_api_id` is the id of the API Gateway.
+
+`resource_id` is the id of the resource `/api/v1/<resource>`.
+
+`http_method` is the HTTP method ie. `GET`.
+
+
+### Adding the integration for the method
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/api_gateway/README.md:94]: No you can add a method without integration. It's the one above `Adding a method (GET /api/v1/<resource>)`

I added some more information.

File: terraform/authentication/env/dev/example_tfvars.txt [hunk 28]
```diff
@@ -0,0 +1,17 @@
...
+  refresh_token = 28
+  access_token  = 1
+  id_token      = 1
+}
+token_validity_units = {
+  refresh_token = "days"
+  access_token  = "days"
+  id_token      = "days"
+}
+callback_urls = [
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/authentication/env/dev/example_tfvars.txt:15]: Set them in example_tfvars. I could predefine these based off what a user may set the final API as.

For instance the API url will be `${var.domain}.${var.domain_extension}` i.e. `tech-audit-tool-api.sdp-dev`.

However, only when the API gateway has been created, we get the URL generated by API gateway.

File: terraform/lambda/variables.tf [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+variable "container_ver" {
+  description = "Container tag"
+  type        = string
+  default     = "v0.0.2"
+
+}
+
+variable "image_tag" {
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/lambda/variables.tf]: Done 👍

File: terraform/lambda/variables.tf (hunk 18 above)
Comment (evisos-ons) [reply terraform/lambda/variables.tf]: Done 👍

File: .github/workflows/deploy_mkdocs.yml
[the remark this answers is not in the document]
Comment (evisos-ons) [reply .github/workflows/deploy_mkdocs.yml:1]: Removed as MkDocs was in other PR, my bad.

File: terraform/storage/main.tf (hunk 19 above)
Comment (evisos-ons) [reply terraform/storage/main.tf:13]: Removed.

File: terraform/storage/variables.tf (hunk 20 above)
Comment (evisos-ons) [reply terraform/storage/variables.tf:52]: Removed.

File: README.md [hunk 30]
```diff
@@ -66,7 +66,7 @@ When in root directory, run the testing command. If you are in `aws_lambda_scrip
 make pytest
 ```
 
-If all tests fail, please relogin [here](https://keh-tech-audit-tool.auth.eu-west-2.amazoncognito.com/login?client_id=dm3289s0tqtsr5qn2qm5i9fql&response_type=code&scope=email+openid+phone&redirect_uri=https://dutwj6q915.execute-api.eu-west-2.amazonaws.com/dev/api/verify) to the Cognito and use a new token.
+If all tests fail, please relogin [here](https://tech-audit-tool-api-sdp-dev.auth.eu-west-2.amazoncognito.com/login?client_id=7hdpes1atnuepl1ffvf21a28r3&response_type=code&scope=email+openid&redirect_uri=https://tech-audit-tool-api.sdp-dev.aws.onsdigital.uk/api/v1/verify) to the Cognito and use a new token.
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply README.md:69]: See above comment^

File: terraform/lambda/variables.tf (hunk 12 above)
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/lambda/variables.tf]: This is not used anymore. It pulls from the state now.

File: README.md [hunk 31]
```diff
@@ -48,7 +48,7 @@ poetry run python3 -m app
 
 This repo utilises PyTest for the testing. Please make sure you have installed dev dependencies before running tests.
 
-To test you need a mock token. Visit this [link](https://keh-tech-audit-tool.auth.eu-west-2.amazoncognito.com/login?client_id=dm3289s0tqtsr5qn2qm5i9fql&response_type=code&scope=email+openid+phone&redirect_uri=https://dutwj6q915.execute-api.eu-west-2.amazonaws.com/dev/api/verify) and copy the `id_token`.
+To test you need a mock token. Visit this [link](https://tech-audit-tool-api-sdp-dev.auth.eu-west-2.amazoncognito.com/login?client_id=7hdpes1atnuepl1ffvf21a28r3&response_type=code&scope=email+openid&redirect_uri=https://tech-audit-tool-api.sdp-dev.aws.onsdigital.uk/api/v1/verify) and copy the `id_token`.
```
[the remark this answers is not in the document]
Comment (gibbardsteve) [reply README.md:51]: In the public readme I think it would be better to remove any information specific to acquiring the token other than to say a token needs to be generated for Cognito and that token used run tests. Linking to an internal confluence page is not good for a public repo. You can create a confluence page with specific details.

File: README.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (gibbardsteve) [reply README.md:69]: As above

File: README.md (hunk 31 above)
[the remark this answers is not in the document]
Comment (evisos-ons) [reply README.md]: Removed link. No need for separate docs page for now.

File: README.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (evisos-ons) [reply README.md]: Removed link.

File: terraform/lambda/main.tf [hunk 32]
```diff
@@ -0,0 +1,251 @@
...
+
+# 2. Attach basic execution policy
+resource "aws_iam_role_policy_attachment" "lambda_basic_execution" {
+  role       = aws_iam_role.lambda_execution_role.name
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
+  depends_on = [aws_iam_role.lambda_execution_role]
+}
+
+# 3. Add ECR policy
+resource "aws_iam_role_policy" "lambda_ecr_policy" {
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply terraform/lambda/main.tf]: Removed it.

File: README.md
[the remark this answers is not in the document]
Comment (evisos-ons) [reply README.md:1]: Removed that section of the README 👍

**Review by gibbardsteve — APPROVED** [review]
Looks good
